android/titanium/src/java/org/appcelerator/titanium/TiApplication.java

```diff
@@ -468,23 +467,158 @@ public void postOnCreate()
+	private void deleteTiTempFiles()
+	{
+		// Create the "trash" directory if it doesn't already exist.
+		File trashDir = tryCreateDir(getTiInternalCacheDir(), "trash");
+
+		// Set up an array of all temp directories to be trashed.
+		File[] dirArray = new File[] {
+			// The "Ti.Filesystem.tempDirectory" folder.
+			getTiTempDir(),
+
+			// The legacy temp folder used before Titanium 9.3.0. Won't exist for new app installations.
+			new File(getCacheDir(), "_tmp")
+		};
+
+		// Trash the above directories.
+		String renameSuffix = "_" + System.currentTimeMillis();
+		for (File nextDir : dirArray) {
+			boolean wasTrashed = true;
+			try {
+				// Move folder under the "trash" folder to be deleted asynchronously.
+				if (nextDir.exists()) {
+					wasTrashed = false;
+					wasTrashed = nextDir.renameTo(new File(trashDir, nextDir.getName() + renameSuffix));
+				}
+			} catch (Exception ex) {
+				Log.e(TAG, "Failed to trash directory: " + nextDir, ex);
+			} finally {
+				// If failed to move existing folder to "trash", then do a blocking delete. (Should never happen.)
+				if (!wasTrashed) {
+					tryDeleteTree(nextDir);
+				}
+			}
 		}
-		return cacheDir.getAbsoluteFile();
+
+		// Async delete the "trash" directory tree.
+		Thread thread = new Thread(() -> {
+			tryDeleteTree(trashDir);
+		});
+		thread.start();
+	}
+
+	/**
+	 * Recursively deletes the given directory tree.
+	 * Will never throw an exception and will return the result as a boolean instead.
+	 * @param file Reference to a file or directory. Can be null.
+	 * @return
+	 * Returns true if successfully deleted all files and folders under given directory tree.
+	 * Returns false if at least 1 deletion failed or if given a null argument.
+	 */
+	private boolean tryDeleteTree(File file)
+	{
+		boolean wasSuccessful = false;
+		try {
+			if (file != null) {
+				wasSuccessful = deleteTree(file);
+			}
+		} catch (Throwable ex) {
+			Log.e(TAG, "Failed to delete directory tree: " + file, ex);
+		}
+		return wasSuccessful;
+	}
+
+	/**
+	 * Recursively deletes the given directory tree.
+	 * @param file Reference to a directory or a single file. Can be null, in which case this method no-ops.
+	 * @return
+	 * Returns true if successfully deleted all files and folders under given directory tree.
+	 * Returns false if at least 1 deletion failed or if given a null argument.
+	 * @exception SecurityException Thrown if don't have permission to delete at least 1 file in the tree.
+	 */
+	private boolean deleteTree(File file) throws SecurityException
+	{
+		// Validate argument.
+		if (file == null) {
+			return false;
+		}
+
+		// If given a directory, then recursively delete the entire tree.
+		boolean wasDeleted = true;
+		if (file.isDirectory()) {
+			for (File nextFile : file.listFiles()) {
+				wasDeleted = deleteTree(nextFile) && wasDeleted;
+			}
+		}
+
+		// Delete the given directory/file.
+		return (file.delete() && wasDeleted);
```

Comment:
shouldn't we avoid trying to delete a folder if we know that a child already failed? (i.e. swap the order here)

Reply:
You're right. I'll change it. Thanks.